File: go/internal/cli/pack.go
package cli

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/nexusriot/ai-houkai/internal/memory"
	"github.com/nexusriot/ai-houkai/internal/timeparse"
	"github.com/spf13/cobra"
)

func newPackCmd() *cobra.Command {
	var budget, maxItems, compressMinGroup int
	var memType, tag, mode, header, format, source, since, until string
	var minImp, compressThreshold float32
	var inclSup, compress bool

	cmd := &cobra.Command{
		Use:   "pack <query>",
		Short: "Assemble the most relevant memories into a token-budgeted context block",
		Long: `Assemble the most relevant memories into a token-budgeted context block.

Ranks with hybrid scoring by default, then greedily packs results until the
token budget is reached. The block is printed to stdout (pipe it into a
prompt); a one-line summary goes to stderr.`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			store := storeFromCtx(cmd.Context())
			sinceTS, _, err := timeparse.Parse(since)
			if err != nil {
				return err
			}
			untilTS, _, err := timeparse.Parse(until)
			if err != nil {
				return err
			}
			result, err := store.RecallPack(cmd.Context(), args[0], memory.PackOpts{
				TokenBudget:       budget,
				Type:              memory.MemoryType(memType),
				Tag:               tag,
				MinImportance:     minImp,
				Mode:              memory.RecallMode(mode),
				MaxItems:          maxItems,
				IncludeSuperseded: inclSup,
				Header:            &header,
				Source:            source,
				Since:             sinceTS,
				Until:             untilTS,
				Compress:          compress,
				CompressThreshold: compressThreshold,
				CompressMinGroup:  compressMinGroup,
			})
			if err != nil {
				return err
			}

			if format == "json" {
				type item struct {
					ID         string   `json:"id"`
					Text       string   `json:"text"`
					Type       string   `json:"type"`
					Tags       []string `json:"tags"`
					Importance float32  `json:"importance"`
					Score      float32  `json:"score"`
					Tokens     int      `json:"tokens"`
				}
				items := make([]item, len(result.Items))
				for i, p := range result.Items {
					items[i] = item{
						ID:         p.Memory.ID,
						Text:       p.Memory.Text,
						Type:       string(p.Memory.Type),
						Tags:       p.Memory.Tags,
						Importance: p.Memory.Importance,
						Score:      p.Score,
						Tokens:     p.Tokens,
					}
				}
				payload := map[string]any{
					"text":        result.Text,
					"used_tokens": result.UsedTokens,
					"budget":      result.Budget,
					"truncated":   result.Truncated,
					"items":       items,
				}
				if len(result.CompressedGroups) > 0 {
					groups := make([]map[string]any, len(result.CompressedGroups))
					for i, g := range result.CompressedGroups {
						groups[i] = map[string]any{
							"ids": g.IDs(), "count": len(g.Memories), "text": g.Text, "tokens": g.Tokens,
						}
					}
					payload["compressed_groups"] = groups
				}
				b, _ := json.MarshalIndent(payload, "", "  ")
				fmt.Println(string(b))
				return nil
			}

			if len(result.Items) == 0 {
				fmt.Fprintln(os.Stderr, "No memories found.")
				return nil
			}

			fmt.Println(result.Text)
			truncNote := ""
			if result.Truncated {
				truncNote = " · truncated"
			}
			fmt.Fprintf(os.Stderr, "[%d memories · %d/%d tokens%s]\n",
				len(result.Items), result.UsedTokens, result.Budget, truncNote)
			return nil
		},
	}
	cmd.Flags().IntVarP(&budget, "budget", "b", 800, "Token budget for the packed block")
	cmd.Flags().StringVarP(&memType, "type", "t", "", "Filter by memory type")
	cmd.Flags().StringVarP(&tag, "tag", "g", "", "Filter by tag")
	cmd.Flags().Float32Var(&minImp, "min-importance", 0, "Minimum importance")
	cmd.Flags().StringVar(&mode, "mode", "hybrid", "semantic|hybrid")
	cmd.Flags().IntVar(&maxItems, "max-items", 50, "Ranked candidates to consider")
	cmd.Flags().BoolVar(&inclSup, "include-superseded", false, "Include superseded memories")
	cmd.Flags().StringVar(&header, "header", "## Relevant memory", "Block header (empty string to omit)")
	cmd.Flags().StringVarP(&format, "output", "f", "text", "Output format: text|json")
	cmd.Flags().StringVar(&source, "source", "", "Filter by exact provenance string")
	cmd.Flags().StringVar(&since, "since", "", "Only memories created at/after (ISO date, epoch, or '7d')")
	cmd.Flags().StringVar(&until, "until", "", "Only memories created at/before (ISO date, epoch, or '7d')")
	cmd.Flags().BoolVar(&compress, "compress", false, "Fold budget-dropped, similar memories into compressed summary lines")
	cmd.Flags().Float32Var(&compressThreshold, "compress-threshold", 0.30, "Jaccard similarity for compression clustering")
	cmd.Flags().IntVar(&compressMinGroup, "compress-min-group", 2, "Minimum cluster size to compress")
	return cmd
}

File: ai_houkai/cli/commands/pack.py
from __future__ import annotations

import json
from typing import Optional

import typer

from ai_houkai.cli import output as out
from ai_houkai.timeparse import parse_timestamp


def pack(
    ctx: typer.Context,
    query: str = typer.Argument(..., help="Semantic search query"),
    budget: int = typer.Option(800, "-b", "--budget", help="Token budget for the packed block"),
    type: Optional[str] = typer.Option(None, "-t", "--type"),
    tag: Optional[str] = typer.Option(None, "-g", "--tag"),
    min_importance: Optional[float] = typer.Option(None, "--min-importance"),
    source: Optional[str] = typer.Option(None, "--source", help="Filter by exact provenance string"),
    since: Optional[str] = typer.Option(None, "--since", help="Only memories created at/after (ISO date, epoch, or '7d')"),
    until: Optional[str] = typer.Option(None, "--until", help="Only memories created at/before (ISO date, epoch, or '7d')"),
    mode: str = typer.Option("hybrid", "--mode", help="semantic|hybrid"),
    max_items: int = typer.Option(50, "--max-items", help="Ranked candidates to consider"),
    include_superseded: bool = typer.Option(False, "--include-superseded"),
    header: str = typer.Option("## Relevant memory", "--header", help="Block header (empty string to omit)"),
    fmt: str = typer.Option("text", "--format", "-f", help="text|json"),
) -> None:
    """Assemble the most relevant memories into a token-budgeted context block.

    Ranks with hybrid scoring by default, then greedily packs results until the
    token budget is reached. The block is printed to stdout (pipe it into a
    prompt); a one-line summary goes to stderr.
    """
    store = ctx.obj["store"]
    try:
        since_ts = parse_timestamp(since)
        until_ts = parse_timestamp(until)
    except ValueError as exc:
        raise typer.BadParameter(str(exc))
    with out.friendly_errors():
        result = store.recall_pack(
            query=query,
            token_budget=budget,
            type=type,
            tag=tag,
            min_importance=min_importance,
            source=source,
            since=since_ts,
            until=until_ts,
            mode=mode,
            max_items=max_items,
            include_superseded=include_superseded,
            header=header,
        )

    if fmt == "json":
        typer.echo(json.dumps(
            {
                "text": result.text,
                "used_tokens": result.used_tokens,
                "budget": result.budget,
                "truncated": result.truncated,
                "items": [
                    {
                        "id": p.memory.id,
                        "text": p.memory.text,
                        "type": p.memory.type,
                        "tags": p.memory.tags,
                        "importance": p.memory.importance,
                        "score": round(p.score, 4),
                        "tokens": p.tokens,
                    }
                    for p in result.items
                ],
            },
            indent=2,
        ))
        return

    if not result.items:
        typer.echo("No memories found.", err=True)
        return

    typer.echo(result.text)
    summary = (
        f"[{len(result.items)} memories · {result.used_tokens}/{result.budget} tokens"
        f"{' · truncated' if result.truncated else ''}]"
    )
    typer.echo(summary, err=True)
